### PR TITLE
Correctness fixes for DFlash training path

### PR DESCRIPTION
**Pull request by dcw02**
## Motivation

Upstreaming some correctness fixes to the DFlash training path:
- Add support for sliding window attention (causal)
- Allow for per layer configuration of the draft model for either sliding or full attention
- Fix eager attention mask construction for both sliding and full attention layers
- Fix anchor sampling so anchors are drawn from valid supervised positions
- Normalize dflash loss globally across distributed ranks and gradient accumulation steps
- Adds regression tests for the updated masks, anchor sampling, and distributed loss normalization

### CODE REVIEW

**Comment by jiapingW** [comment]
LGTM. But you have to fix the unittest that can be solved by 
```
--- a/tests/test_scripts/test_expand_reasoning_conversations.py
+++ b/tests/test_scripts/test_expand_reasoning_conversations.py
@@ -34,7 +34,7 @@ def _load_preprocessing_stack():
     distributed_module.get_sp_ring_group = lambda: None
     sys.modules[f"{package_name}.distributed"] = distributed_module
 
-    for module_name in ("template", "parse", "preprocessing"):
+    for module_name in ("template", "parse", "loss_mask", "preprocessing"):
         full_name = f"{data_package_name}.{module_name}"
         module_path = repo_root / "specforge" / "data" / f"{module_name}.py"
         spec = importlib.util.spec_from_file_location(full_name, module_path)
```

**Comment by jiapingW** [comment]
@dcw02 Can you help fix the unit test?

**Comment by dcw02** [comment]
yes fixing now

**Comment by dcw02** [comment]
@jiapingW fixed

**Comment by jiapingW** [comment]
Your test doesn't seperate the distribution communication. And the ` _reduce_ratio_metrics()` function needs to be updated that skipped the `all_reduce` when `world_size ==1`. @dcw02

**Review by jiapingW — APPROVED** [review]
